File: t/PACSFramework/VB/StorageServerManager/DataTypes/LogAudit/LocalAuditLogQueue.vb
' *************************************************************
' Copyright (c) 1991-2019 LEAD Technologies, Inc.              
' All Rights Reserved.                                         
' *************************************************************
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Text
Imports Leadtools.Logging
Imports Leadtools.Logging.Medical
Imports System.Collections.Specialized

Namespace Leadtools.Demos.StorageServer.DataTypes
   Friend MustInherit Class LocalAuditLogQueue
      Private Shared _audits As OrderedDictionary = New OrderedDictionary()
      Private Shared _lock As Object = New Object()

      Public Shared Sub AddAuditMessage(ByVal id As Integer, ByVal message As String)
         SyncLock _lock
            If _audits.Contains(id) Then
               _audits.Remove(id)
            End If

            _audits.Add(id, message)
         End SyncLock
      End Sub

      Public Shared Sub FlushLogs(ByVal logger As Logger, ByVal user As String)
         SyncLock _lock
            Do While _audits.Count <> 0
               Dim logEntry As DicomLogEntry = New DicomLogEntry()

               logEntry.ClientAETitle = user
               logEntry.Description = _audits(0).ToString()
               logEntry.LogType = LogType.Audit

               logger.Log(logEntry)

               _audits.RemoveAt(0)
            Loop
         End SyncLock
      End Sub

      Public Shared Sub Clear()
         SyncLock _lock
            _audits.Clear()
         End SyncLock
      End Sub
   End Class
End Namespace

File: t/PACSFramework/CS/Leadtools.Medical.Winforms/LogAudit/LocalAuditLogQueue.cs
// *************************************************************
// Copyright (c) 1991-2019 LEAD Technologies, Inc.              
// All Rights Reserved.                                         
// *************************************************************
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using Leadtools.Logging;
using Leadtools.Logging.Medical;
using System.Collections.Specialized;

namespace Leadtools.Medical.Winforms
{
   public abstract class LocalAuditLogQueue
   {
      static OrderedDictionary _audits = new OrderedDictionary ( ) ;
      static object _lock = new object ( ) ;
      
      public static void AddAuditMessage ( int id, string message ) 
      {
         lock ( _lock ) 
         {
            if ( _audits.Contains ( id ) )
            {
               _audits.Remove ( id ) ;
            }
            
            _audits.Add ( id, message ) ;
         }
      }
      
      public static void FlushLogs ( Logger logger, string user ) 
      {
         lock ( _lock ) 
         {
            while ( _audits.Count != 0 )
            {
               DicomLogEntry logEntry = new DicomLogEntry ( ) ;
               
               logEntry.ClientAETitle = user ;
               logEntry.Description   = _audits [ 0 ].ToString ( ) ;
               logEntry.LogType       = LogType.Audit ;
               
               logger.Log ( logEntry ) ;
               
               _audits.RemoveAt ( 0 ) ;
            }
         }
      }
      
      public static void Clear ( ) 
      {
         lock ( _lock ) 
         {
            _audits.Clear ( ) ;
         }
      }
   }
}
